library/src/main/java/com/novoda/downloadmanager/DownloadManager.java

```diff
@@ -123,48 +125,60 @@ public void removeDownloadBatchCallback(DownloadBatchCallback downloadBatchCallb
+    private List<DownloadBatchStatus> executeGetAllDownloadBatchStatuses() {
         List<DownloadBatchStatus> downloadBatchStatuses = new ArrayList<>(downloadBatchMap.size());
 
         for (DownloadBatch downloadBatch : downloadBatchMap.values()) {
             downloadBatchStatuses.add(downloadBatch.status());
         }
+        return downloadBatchStatuses;
+    }
 
-        callbackHandler.post(() -> callback.onReceived(downloadBatchStatuses));
+    @Override
+    public void getAllDownloadBatchStatuses(AllBatchStatusesCallback callback) {
+        executor.submit((Runnable) () -> WaitForDownloadServiceThenPerform.<Void>waitFor(downloadService, waitForDownloadService)
+                .thenPerform(() -> {
+                    List<DownloadBatchStatus> downloadBatchStatuses = executeGetAllDownloadBatchStatuses();
+                    callbackHandler.post(() -> callback.onReceived(downloadBatchStatuses));
+                    return null;
+                }));
     }
 
+    @Nullable
+    @WorkerThread
     @Override
-    public void getDownloadStatusWithMatching(DownloadFileId downloadFileId, DownloadFileStatusCallback callback) {
-        if (downloadService == null) {
-            executor.submit(
-                    WaitForLockRunnable.waitFor(waitForDownloadService)
-                            .thenPerform(() -> executeFirstLocalPathForDownloadWithMatching(downloadFileId, callback))
-            );
-        } else {
-            executeFirstLocalPathForDownloadWithMatching(downloadFileId, callback);
-        }
+    public DownloadFileStatus getDownloadStatusWithMatching(DownloadFileId downloadFileId) {
+        return WaitForDownloadServiceThenPerform.<DownloadFileStatus>waitFor(downloadService, waitForDownloadService)
+                .thenPerform(() -> executeGetDownloadStatusWithMatching(downloadFileId));
     }
 
-    private void executeFirstLocalPathForDownloadWithMatching(DownloadFileId downloadFileId, DownloadFileStatusCallback callback) {
+    @Nullable
+    private DownloadFileStatus executeGetDownloadStatusWithMatching(DownloadFileId downloadFileId) {
         for (DownloadBatch downloadBatch : downloadBatchMap.values()) {
-            DownloadFile downloadFile = downloadBatch.downloadFileWith(downloadFileId);
-            if (downloadFile != null) {
-                callbackHandler.post(() -> callback.onReceived(downloadFile.fileStatus()));
-                return;
+            DownloadFileStatus downloadFileStatus = downloadBatch.downloadFileStatusWith(downloadFileId);
+
+            if (downloadFileStatus != null) {
+                return downloadFileStatus;
             }
         }
+        return null;
+    }
+
+    @Override
+    public void getDownloadStatusWithMatching(DownloadFileId downloadFileId, DownloadFileStatusCallback callback) {
+        executor.submit((Runnable) () -> WaitForDownloadServiceThenPerform.<Void>waitFor(downloadService, waitForDownloadService)
```

Comment:
I can use `Callable` which removes the `Runnable` cast but then I need to remove a warning for an exceptional return value that is not used 😬  so ANNOYING!